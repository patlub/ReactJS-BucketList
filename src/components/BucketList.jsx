import React, {Component} from 'react';
import '../App.css';
import axiosInstance from './config';
import _ from 'lodash';

const bucketListItems = []

class BucketList extends Component {
    constructor(props) {
        super(props);
        this.state = {
            isEditing: false,
            items: false
        };
    }

    renderActions() {
        if (this.state.isEditing) {
            return (
                <td>
                    <button onClick={this.onSaveClick.bind(this)} className="btn btn-sm btn-success action-btn">Save
                    </button>
                    <button onClick={this.onCancelClick.bind(this)} className="btn btn-sm btn-default">Cancel</button>
                </td>
            );
        }
        return (
            <td>
                <button onClick={this.onEditClick.bind(this)} className="btn btn-sm btn-primary action-btn">Edit
                </button>
                <button onClick={this.onDeleteClick.bind(this)} className="btn btn-sm btn-danger">Delete</button>
            </td>
        );
    }

    render_buckets() {
        if (this.state.isEditing) {
            return (
                <tr>
                    <td><input type="text" defaultValue={this.props.name} ref="editBucketName"/></td>
                    <td><input type="text" defaultValue={this.props.desc} ref="editDesc"/></td>
                    <td>{this.props.date_added}</td>
                    <input type="hidden" value={this.props.id} ref="id"/>
                    {this.renderActions()}
                </tr>
            );
        }

        return (
            <tr>
                <td onClick={this.onViewItemClick.bind(this)}>{this.props.name}</td>
                <td>{this.props.desc}</td>
                <td>{this.props.date_added}</td>
                <input type="hidden" value={this.props.id} ref="id"/>
                {this.renderActions()}
            </tr>
        );
    }

    render() {
        return this.render_buckets();
    }

    onEditClick() {
        this.setState({isEditing: true})
    }

    onCancelClick() {
        this.setState({isEditing: false})
    }

    onSaveClick(event) {
        event.preventDefault();
        const bucket = this.refs.editBucketName.value;
        const desc = this.refs.editDesc.value;
        const id = this.refs.id.value;

        axiosInstance.put('/buckets/' + id,
            {
                bucket,
                desc
            })
            .then((response) => {
                this.props.unSetBuckets();
                this.setState({isEditing: false});
                this.props.getBuckets();
            })
            .catch((error) => {
                console.log(error);
            });
    }

    onDeleteClick(event) {
        event.preventDefault();
        const id = this.refs.id.value;

        axiosInstance.delete('/buckets/' + id)
            .then((response) => {
                this.props.unSetBuckets();
                this.setState({isEditing: false});
                this.props.getBuckets();
            })
            .catch((error) => {
                console.log(error);
            });
    }

    onViewItemClick(event){
        axiosInstance.get('/items/' + this.props.id)
        .then(function (response) {
            if (response.status === 200) {
                _.forEach(response.data, function (value) {
                    bucketListItems.push(value);
                });
                // console.log(bucketListItems);
                this.setState({items: bucketListItems});
            }
        }.bind(this))
        .catch(function (error) {
            console.log(error);
        });
        console.log(this.props.id);
        
    }
}
export default BucketList;
