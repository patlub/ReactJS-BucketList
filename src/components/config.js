import axios from 'axios'


const axiosInstance = axios.create({
    baseURL: 'https://patrickluboobi-bucket-list-api.herokuapp.com',
});

export default axiosInstance
/**
 * Created by PATRICK on 8/17/2017.
 */
