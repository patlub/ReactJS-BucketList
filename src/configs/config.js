
/**
 * Created by PATRICK on 8/17/2017.
 */
export const baseURL = 'https://awesome-bucket.herokuapp.com';

