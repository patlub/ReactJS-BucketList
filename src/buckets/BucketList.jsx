import React, {Component} from 'react';

// styles
import '../App.css';

// configs
import {baseURL} from '../configs/config';

// Third party
import axios from 'axios'
import {NotificationContainer, NotificationManager} from 'react-notifications';
import _ from 'lodash';

class BucketList extends Component {
    constructor(props) {
        super(props);
        this.state = {
            isEditing: false,
            bucket: this.props.name,
            desc: this.props.desc,
        };
    }

    resetState = () => {
        this.setState({
            isEditing: false,
            bucket: this.props.name,
            desc: this.props.desc,
        })
    };

    /*
    * Fired when input changes
    * @param (event) event when input changes
    * */
    onInputChanged = (event) => {
        this.setState({
            [event.target.name]: event.target.value
        })
    };

    /**
     * This method handles viewing a bucket.
     */
    onViewBucketClick = () => {
        const bucketListItems = [];
        axios.defaults.headers.common['Authorization'] = localStorage.getItem('token');
        axios.get(`${baseURL}/items/${this.props.id}`)
            .then(function (response) {
                _.forEach(response.data, function (value) {
                    bucketListItems.push(value);
                });
                this.props.getItems(bucketListItems, this.props.id);
            }.bind(this))
            .catch(function (error) {
                console.log(error);
            });
    };

    /**
     * This method handles saving an updated a bucket.
     * @param {event} event when triggering an onSaveClick action.
     */
    onSaveClick = (event) => {
        event.preventDefault();
        axios.defaults.headers.common['Authorization'] = localStorage.getItem('token');
        axios.put(`${baseURL}/buckets/${this.props.id}`,
            {
                bucket: this.state.bucket,
                desc: this.state.desc,
            })
            .then(() => {
                this.setState({isEditing: false});
                this.props.updateBuckets(this.props.id, this.state.bucket, this.state.desc);
                NotificationManager.success(`Bucket updated`, `Success`);
            })
            .catch(() => {
                NotificationManager.error(`Bucket ${this.state.bucket} already exists`, `Error`);
            });
    };

    /**
     * This method handles deleting a bucket.
     */
    onDeleteClick = (event) => {
        event.preventDefault();
        axios.defaults.headers.common['Authorization'] = localStorage.getItem('token');
        axios.delete(`${baseURL}/buckets/${this.props.id}`)
            .then(() => {
                this.setState({isEditing: false});
                this.props.deleteBucket(this.props.id);
                NotificationManager.success(`Bucket Deleted`, `Deleted`);
            })
            .catch((error) => {
                NotificationManager.error(`Could not delete bucket`, `Error`);
            });
    };

    onEditClick = () => {
        this.setState({isEditing: true})
    };

    onCancelClick = () => {
        this.resetState();
    };

    // Displays the actions operable on a bucket
    renderActions = () => {
        if (this.state.isEditing) {
            return (
                <td>
                    <span onClick={this.onSaveClick} className="glyphicon glyphicon-floppy-disk save"></span>
                    <span onClick={this.onCancelClick} className="glyphicon glyphicon-ban-circle cancel"></span>

                </td>
            );
        }
        return (
            <td>
                <span onClick={this.onEditClick} className="glyphicon glyphicon-pencil edit"></span>
                <span onClick={this.onDeleteClick} className="glyphicon glyphicon-trash remove"></span>
            </td>
        );
    };

    // Displays the buckets
    renderBuckets = () => {
        if (this.state.isEditing) {
            return this.editingState();
        }
        return (
            <tr>
                <td onClick={this.onViewBucketClick} className="bucket-name">{this.state.bucket}</td>
                <td>{this.state.desc}</td>
                <td>{this.props.date_added}</td>
                {this.renderActions()}
                <NotificationContainer/>
            </tr>
        );
    };

    // Rendered when state isEditing is true
    editingState = () => (
        <tr>
            <td>
                <input
                    type="text"
                    name="bucket"
                    value={this.state.bucket}
                    onChange={this.onInputChanged}/>
            </td>
            <td>
                <input
                    type="text"
                    name="desc"
                    value={this.state.desc}
                    onChange={this.onInputChanged}/>
            </td>
            <td>{this.props.date_added}</td>
            {this.renderActions()}
        </tr>
    );

    render() {
        return this.renderBuckets();
    }
}

export default BucketList;
