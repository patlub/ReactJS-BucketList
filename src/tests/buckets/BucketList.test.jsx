import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import BucketList from '../../buckets/BucketList';
import { baseURL } from '../../configs/config';

jest.mock('react-notifications');

const axios = require('axios');
const MockAdapter = require('axios-mock-adapter');

// This sets the mock adapter on the default instance
const mock = new MockAdapter(axios);

// Mock update bucket PUT request to /buckets/1
mock.onPut(`${baseURL}/buckets/1`).reply(200, {
});

// Mock delete bucket DELETE request to /buckets/1
mock.onDelete(`${baseURL}/buckets/1`).reply(200, {
});

// Mock get items GET request to /items/1
mock.onGet(`${baseURL}/items/1`).reply(200, {
});

global.localStorage = {
  getItem: () => {},
  setItem: () => {},
};

global.confirm = () => true;

describe('Component: Login', () => {
  const props = {
    id: 1,
    name: 'Travel',
    desc: 'cities',
    updateBuckets: () => {},
    deleteBuckets: () => {},
    getItems: () => {},
  };
  const bucketListComponent = mount(<BucketList {...props} />);
  const bucketName = bucketListComponent.find('td.main-name');
  const editIconBtn = bucketListComponent.find('span.edit');
  const deleteIconBtn = bucketListComponent.find('span.remove');

  it('Displays BucketList component', () => {
    const rendered = renderer.create(
      <BucketList />,
    );
    expect(rendered.toJSON()).toMatchSnapshot();
  });
  it('sets props to state', () => {
    expect(bucketListComponent.state().isEditing).toEqual(false);
    expect(bucketListComponent.state().bucket).toEqual('Travel');
    expect(bucketListComponent.state().desc).toEqual('cities');
  });
  it('sets isEditing state to true', () => {
    editIconBtn.simulate('click');
    expect(bucketListComponent.state().isEditing).toEqual(true);
  });
  it('sets isEditing state to true', () => {
    editIconBtn.simulate('click');
    const cancelIconBtn = bucketListComponent.find('span.cancel');
    cancelIconBtn.simulate('click');
    expect(bucketListComponent.state().isEditing).toEqual(false);
  });
  it('sets isEditing state to false', () => {
    editIconBtn.simulate('click');
    const cancelIconBtn = bucketListComponent.find('span.cancel');
    cancelIconBtn.simulate('click');
    expect(bucketListComponent.state().isEditing).toEqual(false);
  });
  it('sets state on input changed', () => {
    editIconBtn.simulate('click');
    const bucketNameInput = bucketListComponent.find('[type="text"]').at(0);
    const bucketDescInput = bucketListComponent.find('[type="text"]').at(1);
    bucketNameInput.simulate('change', { target: { name: 'bucket', value: 'Food' } });
    bucketDescInput.simulate('change', { target: { name: 'desc', value: 'crabs' } });
    expect(bucketListComponent.state().bucket).toEqual('Food');
    expect(bucketListComponent.state().desc).toEqual('crabs');
  });
  it('saves on click save', () => {
    editIconBtn.simulate('click');
    const saveIconBtn = bucketListComponent.find('span.save');
    saveIconBtn.simulate('click');
  });
  it('deletes on click delete', () => {
    deleteIconBtn.simulate('click');
  });
  it('fetches items on click bucket', () => {
    bucketName.simulate('click');
  });
});
