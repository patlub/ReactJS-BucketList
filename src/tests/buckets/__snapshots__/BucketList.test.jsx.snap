// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Displays BucketList component 1`] = `
<tr>
  <td
    className="bucket-name"
    onClick={[Function]}
  />
  <td />
  <td />
  <td>
    <span
      className="glyphicon glyphicon-pencil edit"
      onClick={[Function]}
    />
    <span
      className="glyphicon glyphicon-trash remove"
      onClick={[Function]}
    />
  </td>
  <div
    className="notification-container notification-container-empty"
  >
    <span />
  </div>
</tr>
`;
